### PR TITLE
Write ugrid dataset group

### PR DESCRIPTION
**Pull request by vcloarec**
This PR add the possibility to persist dataset groups for 2D mesh UGRID format

### CODE REVIEW

File: mdal/frmts/mdal_ugrid.cpp [hunk 1]
```diff
@@ -695,65 +702,275 @@ void MDAL::DriverUgrid::save( const std::string &fileName, const std::string &,
...
+  try
+  {
+    mNcFile->openFile( fileName, true );
+  }
+  catch ( MDAL::Error &err )
+  {
+    err.setDriver( name() );
+    MDAL::Log::error( err );
+    return true;
+  }
```
Comment (PeterPetrik) [inline mdal/frmts/mdal_ugrid.cpp:732]: I would normally assume that you open file once in the persist function, add all groups/variables and then close a file. Why you need to open/close files for each dataset group? is it consistent with other drivers?

File: mdal/frmts/mdal_ugrid.cpp [hunk 2]
```diff
@@ -782,10 +999,6 @@ void MDAL::DriverUgrid::writeVariables( MDAL::Mesh *mesh )
     }
   }
 
-  // Time array
-  int timeId = mNcFile->defineVar( "time", NC_DOUBLE, 1, &dimTimeId );
-  mNcFile->putAttrStr( timeId, "units", "hours since 2000-01-01 00:00:00" );
-
```
Comment (PeterPetrik) [inline mdal/frmts/mdal_ugrid.cpp:788]: why this is removed?

File: mdal/frmts/mdal_ugrid.cpp [hunk 3]
```diff
@@ -878,12 +1091,6 @@ void MDAL::DriverUgrid::writeVariables( MDAL::Mesh *mesh )
       faceIndex += facesRead;
     }
   }
-
-  // Time values (not implemented)
-  mNcFile->putDataDouble( timeId, 0, 0.0 );
-
-  // Turning on define mode
-  nc_redef( mNcFile->handle() );
```
Comment (PeterPetrik) [inline mdal/frmts/mdal_ugrid.cpp:886]: why this is removed?

File: tests/test_ugrid.cpp [hunk 4]
```diff
@@ -814,6 +814,306 @@ TEST( MeshUgridTest, flow_3d )
...
+  MDAL_CloseMesh( m );
+
+  m = MDAL_LoadMesh( tmpFile.c_str() );
+  ASSERT_NE( m, nullptr );
+  s = MDAL_LastStatus();
+  EXPECT_EQ( MDAL_Status::None, s );
+
+  groupCount = MDAL_M_datasetGroupCount( m );
+  ASSERT_EQ( 14, groupCount );
+
```
Comment (PeterPetrik) [inline tests/test_ugrid.cpp:966]: would be good to check at least one scalar, one vector value on various data locations

File: tests/test_ugrid.cpp [hunk 5]
```diff
@@ -814,6 +814,306 @@ TEST( MeshUgridTest, flow_3d )
...
+  EXPECT_TRUE( success );
+
+  createNewDatasetGroupOnNewFile( m, tmpFile3, refTime, timeSteps, "new group vector faces", MDAL_DataLocation::DataOnFaces, faceCount, false, 1.23, success );
+  EXPECT_TRUE( success );
+
+  createNewDatasetGroupOnNewFile( m, tmpFile4, refTime, timeSteps, "new group vector vertices", MDAL_DataLocation::DataOnVertices, vertexCount, false, 4.56, success );
+  EXPECT_TRUE( success );
+
+  MDAL_CloseMesh( m );
+
```
Comment (PeterPetrik) [inline tests/test_ugrid.cpp:1110]: would be good to load the written file and check few values/variables?

File: mdal/frmts/mdal_ugrid.cpp (hunk 1 above)
Comment (vcloarec) [reply mdal/frmts/mdal_ugrid.cpp:732]: Same logic for other drivers. Persisting dataset group concerns only one dataset group.

File: mdal/frmts/mdal_ugrid.cpp (hunk 2 above)
Comment (vcloarec) [reply mdal/frmts/mdal_ugrid.cpp:788]: This method is called only for writing mesh frame. In that case, there is no time, and I think UGRID supports no time. Further more, settings a dummy time here leads to not be able to check if the time is the same when writing new dataset group.
The logic is:
- if there is already times in the file, MDAL checks if the times (ref and time steps) are the same, error if not
- if there is no times, adds time depending of the times of the new dataset group.

File: mdal/frmts/mdal_ugrid.cpp (hunk 1 above)
Comment (PeterPetrik) [reply mdal/frmts/mdal_ugrid.cpp:732]: ok cool

File: mdal/frmts/mdal_ugrid.cpp (hunk 3 above)
Comment (vcloarec) [reply mdal/frmts/mdal_ugrid.cpp:886]: for times, same reason as above.

For define mode, I think it is an old left over. Define mode has to be set before adding new variables, not at the end.

File: tests/test_ugrid.cpp (hunk 4 above)
Comment (vcloarec) [reply tests/test_ugrid.cpp:966]: dataset values are checked in `createNewDatasetGroupOnExistingFile()`

File: tests/test_ugrid.cpp (hunk 5 above)
Comment (vcloarec) [reply tests/test_ugrid.cpp:1110]: dataset values are checked in `createNewDatasetGroupOnNewFile()`

**Pull request merged**